lib/crypto/src/poseidon2/mod.rs

```diff
@@ -0,0 +1,294 @@
+#[derive(Clone, Copy, Debug)]
+enum Mode {
+    Absorbing,
+    Squeezing,
+}
+
+/// Poseidon2 sponge that can absorb any number of [`F`] field elements and be
+/// squeezed to a finite number of [`F`] field elements.
+#[derive(Clone, Debug)]
+pub struct Poseidon2<P: PoseidonParams<F>, F: PrimeField> {
+    phantom: core::marker::PhantomData<P>,
+    state: Box<[F]>,
+    mode: Mode,
+    index: usize,
+}
+
+impl<P: PoseidonParams<F>, F: PrimeField> Default for Poseidon2<P, F> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<P: PoseidonParams<F>, F: PrimeField> Poseidon2<P, F> {
+    /// Create a new Poseidon sponge.
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            phantom: core::marker::PhantomData,
+            state: vec![F::zero(); P::T].into_boxed_slice(),
+            mode: Mode::Absorbing,
+            // Begin index from `CAPACITY`. Skip capacity elements.
+            index: P::CAPACITY,
+        }
+    }
+
+    /// Size of poseidon sponge's state.
+    #[must_use]
+    pub const fn state_size() -> usize {
+        P::T
+    }
+
+    /// Start index of partial rounds.
+    #[must_use]
+    const fn partial_round_start() -> usize {
+        P::ROUNDS_F / 2
+    }
+
+    /// End index of partial rounds (noninclusive).
+    #[must_use]
+    const fn partial_round_end() -> usize {
+        Self::partial_round_start() + P::ROUNDS_P
+    }
+
+    /// Total number of rounds.
+    #[must_use]
+    const fn rounds() -> usize {
+        P::ROUNDS_F + P::ROUNDS_P
+    }
+
+    /// Absorb a single element into the sponge.
+    ///
+    /// # Panics
+    ///
+    /// May panic if absorbing while squeezing.
+    pub fn absorb(&mut self, elem: &F) {
+        if let Mode::Squeezing = self.mode {
+            panic!("cannot absorb while squeezing");
+        }
+
+        if self.index == Self::state_size() {
+            self.permute();
+            self.index = P::CAPACITY;
+        }
+
+        self.state[self.index] += elem;
+        self.index += 1;
+    }
+
+    /// Absorb batch of elements into the sponge.
+    pub fn absorb_batch(&mut self, elems: &[F]) {
+        for elem in elems {
+            self.absorb(elem);
+        }
+    }
+
+    /// Permute elements in the sponge.
+    pub fn permute(&mut self) {
+        // Linear layer at the beginning.
+        self.matmul_external();
+
+        // Run the first half of the full round.
+        for round in 0..Self::partial_round_start() {
+            self.add_rc_external(round);
+            self.apply_sbox_external();
+            self.matmul_external();
+        }
+
+        // Run the partial round.
+        for round in Self::partial_round_start()..Self::partial_round_end() {
+            self.add_rc_internal(round);
+            self.apply_sbox_internal();
+            self.matmul_internal();
+        }
+
+        // Run the second half of the full round.
+        for round in Self::partial_round_end()..Self::rounds() {
```

Comment:
Same code as for first half of the full rund. Wdyt of having inline fn?

Reply:
Good! Changed it